cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.md

```diff
@@ -107,6 +107,38 @@ csrfToken = hmac + "." + randomValue // Add the `randomValue` to the HMAC hash t
 response.setCookie("csrf_token=" + csrfToken + "; Secure") // Set Cookie without HttpOnly flag
 ```
 
+And the following code demonstrates validation of the CSRF token that is sent back from the client:
+
+```code
+// Get the CSRF token from the request
+csrfToken = request.getParameter("csrf_token") // From form field or header
+
+// Split the token to get the randomValue
+const tokenParts = csrfToken.split(".");
+const hmacFromRequest = tokenParts[0];
+const randomValue = tokenParts[1];
+
+// Recreate the HMAC with the current session and the randomValue from the request
+secret = readEnvironmentVariable("CSRF_SECRET") // HMAC secret key
+sessionID = session.sessionID // Current authenticated user session
+message = sessionID.length + "!" + sessionID + "!" + randomValue.length + "!" + randomValue
+
+// Generate the expected HMAC
+expectedHmac = hmac("SHA256", secret, message)
```

Comment:
HMac-SHA256 is going to be expected to return a byte array, not a string, so for here and on line 103, I would recommend changing the call from "hmac" to something like (say) "hmacAsHexString". If you don't do that, then someone trying to implement the pseudo-code on line 104, where it has:

```
   csrfToken = hmac + "." + randomValue;
```

is likely to come out with a mangled string.  Alternate;y, you could do something like:

```
   csrfToken = hmac.asHex() + "." + randomValue.asHex();
```

here. That would save from having to base64-encoding the cookie value.

Reply:
I think those types of implementation details don't belong in pseudocode. The existing pseudocode gives return values the same treatment.

Reply:
Depends. May be true if you do think that no one will feed the pseudo-code to a generative AI. But I don't believe that for a minute.